sw/device/boot_rom2/final_jump.S

```diff
@@ -0,0 +1,91 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+  /**
+   * Linker-provided symbols for stack layout information.
+   */
+  .extern _stack_start
+  .extern _stack_size
+
+_final_jump_asm:
+  .globl _final_jump_asm
+
+  // Bleach the ROM stack, which we no longer need for anything.
+  // Here, |t0| is the current word being zeroed, while |t1| is
+  // the number of bytes left to zero.
+  la   t0, _stack_start
+  la   t1, _stack_size
+1:
+  addi t0, t0, -4
+  addi t1, t1, -4
+  sw   zero, 0(t0)
+  bnez t1, 1b
+
+  // Clobber every register, including |ra| and |sp|. However, do
+  // not clobber |a0| and |a1|, which contain important information.
+  li x1,  0  // ra
+  li x2,  0  // sp
+  li x3,  0
+  li x4,  0
+  li x5,  0
+  li x6,  0
+  li x7,  0
+  li x8,  0
+  li x9,  0
+  // Skip |x10| and |x11|, aka |a0| and |a1|.
+  li x12, 0
+  li x13, 0
+  li x14, 0
+  li x15, 0
+  li x16, 0
+  li x17, 0
+  li x18, 0
+  li x19, 0
+  li x20, 0
+  li x21, 0
+  li x22, 0
+  li x23, 0
+  li x24, 0
+  li x25, 0
+  li x26, 0
+  li x27, 0
+  li x28, 0
+  li x29, 0
+  li x30, 0
+  li x31, 0
+
+  // Copy the final trampoline to the top of the next stage's stack.
+  // This will dirty some registers, but this is non-secret information,
+  // so we won't be re-clobbering them.
```

Comment:
It seems, at first glance, that there could be security implications with executing instructions directly from the stack. That said, I realise it's needed so that we can protect the ROM from being read by flash applications, without trusting the flash applications. I'm not sure of a better solution.

Reply:
This is the approach that me and Garret have agreed to for now, but it still requires some security review. The only other solution I can think of is playing silly games with executable layout and how the ROM is actually locked down as unreadable, but that's more complicated and perhaps not a *better* solution.